From david-b@pacbell.net Fri Dec  8 12:50:41 2006
From: Wojtek Kaniewski <wojtekka@toxygen.net>
Date: Friday 08 December 2006 3:26 am
Subject: USB: at91_udc: Additional checks
To: linux-usb-devel@lists.sourceforge.net
Cc: andrew@sanpeople.com
Message-Id: <200612080942.19025.david-b@pacbell.net>

This patch performs additional checks in at91_udc, just in case of
some spurious interrupts or device enumeration.

Signed-off-by: Wojtek Kaniewski <wojtekka@toxygen.net>
Acked-by: David Brownell <dbrownell@users.sourceforge.net>
Cc: Andrew Victor <andrew@sanpeople.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/at91_udc.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- gregkh-2.6.orig/drivers/usb/gadget/at91_udc.c
+++ gregkh-2.6/drivers/usb/gadget/at91_udc.c
@@ -955,7 +955,10 @@ static int at91_vbus_session(struct usb_
 	// VDBG("vbus %s\n", is_active ? "on" : "off");
 	local_irq_save(flags);
 	udc->vbus = (is_active != 0);
-	pullup(udc, is_active);
+	if (udc->driver)
+		pullup(udc, is_active);
+	else
+		pullup(udc, 0);
 	local_irq_restore(flags);
 	return 0;
 }
@@ -1241,7 +1244,10 @@ static void handle_setup(struct at91_udc
 #undef w_length
 
 	/* pass request up to the gadget driver */
-	status = udc->driver->setup(&udc->gadget, &pkt.r);
+	if (udc->driver)
+		status = udc->driver->setup(&udc->gadget, &pkt.r);
+	else
+		status = -ENODEV;
 	if (status < 0) {
 stall:
 		VDBG("req %02x.%02x protocol STALL; stat %d\n",
